py-polars/polars/internals/frame.py

```diff
@@ -6224,85 +6166,6 @@ def __init__(
         self.by = by
         self.maintain_order = maintain_order
 
-    def __getitem__(self, item: Any) -> GBSelection[DF]:
-        print(
-            "accessing GroupBy by index is deprecated, consider using the `.agg` method"
-        )
-        return self._select(item)
-
-    def _select(self, columns: str | list[str]) -> GBSelection[DF]:  # pragma: no cover
-        """
-        Select the columns that will be aggregated.
-
-        Parameters
-        ----------
-        columns
-            One or multiple columns.
-
-        """
-        warnings.warn(
-            "accessing GroupBy by index is deprecated, consider using the `.agg`"
-            " method",
-            DeprecationWarning,
-        )
-        if isinstance(columns, str):
-            columns = [columns]
-        return GBSelection(
-            self._df,
-            self.by,
-            columns,
-            dataframe_class=self._dataframe_class,
-        )
-
-    def __iter__(self) -> Iterable[Any]:
-        groups_df = self.groups()
-        groups = groups_df["groups"]
-        df = self._dataframe_class._from_pydf(self._df)
-        for i in range(groups_df.height):
-            yield df[groups[i]]
```

Comment:
I see this implementation relies on `self.groups()`, which is deprecated and should be removed. But in general, are we against iterating over a group-by? I've found this functionality helpful every now and then

Reply:
You're right - after thinking about this, it's not necessary to get rid of this functionality entirely.

We could rewrite this without indexing, but I can't seem to come up with something efficient. Suggestions are welcome.

Reply:
This works. Not sure about the time/space complexity...

```python
    def __iter__(self) -> Iterable[Any]:
        groups = self.agg(pli.all())
        df = self._dataframe_class._from_pydf(self._df)
        by = self.by if isinstance(self.by, list) else [self.by]
        non_index_columns = [c for c in df.columns if c not in by]

        for i in range(len(groups)):
            yield groups[i].explode(columns=non_index_columns)
```

Reply:
Another way I thought about approaching this was to first get all the group 'indices' using:

```python
group_indices = df.select(self.by).unique(self.by)
```

And then filter somehow. But I got stuck.

Reply:
Let's solve this some other time! For now, `__iter__` will stay in its current form.